pandas/_libs/lib.pyi

```diff
@@ -38,13 +41,13 @@ def infer_dtype(value: object, skipna: bool = ...) -> str: ...
 def is_iterator(obj: object) -> bool: ...
 def is_scalar(val: object) -> bool: ...
 def is_list_like(obj: object, allow_sets: bool = ...) -> bool: ...
-def is_period(val: object) -> bool: ...
-def is_interval(val: object) -> bool: ...
-def is_decimal(val: object) -> bool: ...
-def is_complex(val: object) -> bool: ...
-def is_bool(val: object) -> bool: ...
-def is_integer(val: object) -> bool: ...
-def is_float(val: object) -> bool: ...
+def is_period(val: object) -> TypeGuard[Period]: ...
+def is_interval(val: object) -> TypeGuard[Interval]: ...
+def is_decimal(val: object) -> TypeGuard[Decimal]: ...
+def is_complex(val: object) -> TypeGuard[complex | np.complex_]: ...
+def is_bool(val: object) -> TypeGuard[bool | np.bool_]: ...
+def is_integer(val: object) -> TypeGuard[int | np.integer]: ...
```

Comment:
I kind of dislike having composite types inside the type guard (because it gives some uglier code in other locations). Maybe only use `TypeGuard` when it gives a single type?

Reply:
i get where youre coming from, but i also have an urge for "maximum accuracy" here.  im going to keep it how it is for now but will be OK with changing if consensus develops

Reply:
I would keep the Union (but it would be nicer without it). I think not having the Unions might cause some incorrect unreachable code warnings from mypy/pyright.

```py
if is_integer(x):
    if isinstance(x, np.number): # I believe mypy would error here if it wouldn't return a Union
        x = x.item()
```

Reply:
I think if we want to check that it is a python integer, we'd just do `isinstance(x, int)`. My issue was more that by having a union type here, we'd have to have this usion type in a lot of places, which doesn't feel like very clean code. My suggestion was to not type `is_integer` or other with union return types.

But ok, it was just a preference, I can also accept it, as overall this is PR is a nice improvement.

Reply:
> My issue was more that by having a union type here, we'd have to have this usion type in a lot of places, which doesn't feel like very clean code.

Or for every place that such a change is necessary, it makes you analyze the code and determine whether you really want to change the test to `isinstance(x, int)` or modify other parts of the code knowing that a  `np.integer` is valid.